### PR TITLE
#NGG-80|resolved| Shared e2e test if there are no focused tests

### PR DESCRIPTION
**Pull request by felixmosh**

### CODE REVIEW

File: protractor-conf.js [hunk 1]
```diff
@@ -23,6 +26,29 @@ config.capabilities = {
...
+      if (!found) {
+        var fileContent = fs.readFileSync(file, 'utf-8');
+        fileContent = fileContent.replace(commentsRegex, '');
+        strings.forEach(function (str) {
+          found = found || fileContent.indexOf(str) >= 0;
+        });
+      }
+    });
+  });
+  return found;
```
Comment (shahata) [inline protractor-conf.js]: You can get rid of this `found` ugliness using `some`:
```js
return patterns.some(function (pattern) {
  return glob.sync(pattern).some(function (file) {
    var fileContent = fs.readFileSync(file, 'utf-8');
    fileContent = fileContent.replace(commentsRegex, '');
    return strings.some(function (str) {
      return fileContent.indexOf(str) >= 0;
    });
  });
});
```

File: protractor-conf.js [hunk 2]
```diff
@@ -23,6 +26,29 @@ config.capabilities = {
...
+}
+
+if(!hasFocusedTests(config.specs, ['iit(', 'ddescribe('])) {
```
Comment (shahata) [inline protractor-conf.js]: add also `fit` and `fdescribe` (jasmine 2.0 is coming :))

File: protractor-conf.js [hunk 3]
```diff
@@ -23,6 +26,29 @@ config.capabilities = {
...
 
+function hasFocusedTests(patterns, strings) {
+  var commentsRegex = /(?:\/\*(?:[\s\S]*?)\*\/)|(?:([\s;])+\/\/(?:.*)$)/gm;
+  var found = false;
+  patterns.forEach(function (pattern) {
+    glob.sync(pattern).forEach(function (file) {
+      if (!found) {
+        var fileContent = fs.readFileSync(file, 'utf-8');
+        fileContent = fileContent.replace(commentsRegex, '');
+        strings.forEach(function (str) {
```
Comment (shahata) [inline protractor-conf.js]: instead of matching against all string I'd use a regexp:
`/(^|\s*)(iit|fit|ddescribe|fdescribe)\(/`

Comment (shahata) [inline protractor-conf.js]: also if you use this regexp, I think you no longer need the comment regexp since we are testing that there is only whitespace before invocation

File: protractor-conf.js [hunk 4]
```diff
@@ -23,6 +26,29 @@ config.capabilities = {
...
+    });
+  });
+  return found;
+}
+
+if(!hasFocusedTests(config.specs, ['iit(', 'ddescribe('])) {
+  // Add capabilities
+  config.capabilities.shardTestFiles = true;
+  config.capabilities.maxInstances = 6;
+}
```
Comment (shahata) [inline protractor-conf.js]: we need to console.log in else so ppl we notice that we opted not to shared

**Comment by felixmosh** [comment]
Changed, thanx for the advice

File: protractor-conf.js [hunk 5]
```diff
@@ -23,6 +26,26 @@ config.capabilities = {
...
+      var fileContent = fs.readFileSync(file, 'utf-8');
+      fileContent = fileContent.replace(commentsRegex, '');
+      return strings.some(function (str) {
+        return fileContent.indexOf(str) >= 0;
+      });
+    });
+  });
+}
+
+if(!hasFocusedTests(config.specs, ['iit(', 'ddescribe('])) {
```
Comment (shahata) [inline protractor-conf.js]: use regexp instead of array: `/^\s*(iit|fit|ddescribe|fdescribe)\(/`

File: protractor-conf.js [hunk 6]
```diff
@@ -23,6 +26,26 @@ config.capabilities = {
   browserName: 'chrome'
 };
 
+function hasFocusedTests(patterns, strings) {
+  var commentsRegex = /(?:\/\*(?:[\s\S]*?)\*\/)|(?:([\s;])+\/\/(?:.*)$)/gm;
+
+  return patterns.some(function (pattern) {
+    return glob.sync(pattern).some(function (file) {
+      var fileContent = fs.readFileSync(file, 'utf-8');
+      fileContent = fileContent.replace(commentsRegex, '');
```
Comment (shahata) [inline protractor-conf.js]: no need to remove comments thanks to regexp

File: protractor-conf.js [hunk 7]
```diff
@@ -23,6 +26,26 @@ config.capabilities = {
...
+      });
+    });
+  });
+}
+
+if(!hasFocusedTests(config.specs, ['iit(', 'ddescribe('])) {
+  // Add capabilities
+  config.capabilities.shardTestFiles = true;
+  config.capabilities.maxInstances = 6;
+}
```
Comment (shahata) [inline protractor-conf.js]: if we are not going to shard, I prefer you add a console log that tells the user something like:
"protractor sharding is disabled due to presence of focused tests"

Comment (shahata) [inline protractor-conf.js]: (make it yellow)

**Comment by shahata** [comment]
merged
